pkg/oidc/provider/authorize.go

```diff
@@ -0,0 +1,239 @@
+package provider
+
+import (
+	"fmt"
+	"net/http"
+	"net/url"
+	"slices"
+	"strings"
+	"time"
+
+	v3 "github.com/rancher/rancher/pkg/apis/management.cattle.io/v3"
+	"github.com/rancher/rancher/pkg/auth/providers"
+	"github.com/rancher/rancher/pkg/auth/tokens"
+	wrangmgmtv3 "github.com/rancher/rancher/pkg/generated/controllers/management.cattle.io/v3"
+	oidcerror "github.com/rancher/rancher/pkg/oidc/provider/error"
+	"github.com/rancher/rancher/pkg/oidc/provider/session"
+	"github.com/rancher/rancher/pkg/settings"
+	"github.com/sirupsen/logrus"
+)
+
+const (
+	supportedResponseType        = "code"
+	supportedCodeChallengeMethod = "S256"
+)
+
+type authParams struct {
+	clientID            string
+	responseType        string
+	scopes              []string
+	codeChallenge       string
+	codeChallengeMethod string
+	nonce               string
+	state               string
+	redirectURI         string
+}
+
+type codeCreator interface {
+	GenerateCode() (string, error)
+}
+
+type sessionAdder interface {
+	Add(code string, session session.Session) error
+}
+
+type authorizeHandler struct {
+	tokenCache      wrangmgmtv3.TokenCache
+	userLister      wrangmgmtv3.UserCache
+	oidcClientCache wrangmgmtv3.OIDCClientCache
+	sessionAdder    sessionAdder
+	codeCreator     codeCreator
+	now             func() time.Time
+}
+
+func newAuthorizeHandler(tokenCache wrangmgmtv3.TokenCache, userLister wrangmgmtv3.UserCache, sessionAdder sessionAdder, codeCreator codeCreator, oidcClientCache wrangmgmtv3.OIDCClientCache) *authorizeHandler {
+	return &authorizeHandler{
+		tokenCache:      tokenCache,
+		userLister:      userLister,
+		sessionAdder:    sessionAdder,
+		codeCreator:     codeCreator,
+		oidcClientCache: oidcClientCache,
+		now:             time.Now,
+	}
+}
+
+func (h *authorizeHandler) authEndpoint(w http.ResponseWriter, r *http.Request) {
+	params, err := getAuthParamsFromRequest(r)
+	if err != nil {
+		oidcerror.WriteError(oidcerror.InvalidRequest, fmt.Sprintf("error parsing parameters from request %v", err), http.StatusBadRequest, w)
+		return
+	}
+	// validate all parameter as per the oidc spec.
+	if params.redirectURI == "" {
+		oidcerror.WriteError(oidcerror.InvalidRequest, "missing redirect_uri", http.StatusBadRequest, w)
+		return
+	}
+	if _, err := url.Parse(params.redirectURI); err != nil {
+		oidcerror.WriteError(oidcerror.InvalidRequest, "invalid redirect_uri", http.StatusBadRequest, w)
+	}
+	if params.responseType != supportedResponseType {
+		oidcerror.RedirectWithError(params.redirectURI, oidcerror.UnsupportedResponseType, "response type not supported", params.state, w, r)
+		return
+	}
+	if params.codeChallengeMethod != supportedCodeChallengeMethod {
+		oidcerror.RedirectWithError(params.redirectURI, oidcerror.InvalidRequest, "challenge_method not supported, only S256 is supported", params.state, w, r)
+		return
+	}
+	if !slices.Contains(params.scopes, "openid") {
```

Comment:
Should this verify that only the "supported scopes" are in this list?

Reply:
done in https://github.com/rancher/rancher/pull/49558/commits/1664be71bb71d75f9e06e13ffab62249ebb6ba40